### PR TITLE
feat(core/pluralize): add singularize capability

### LINKED ISSUE #1743: (core/pluralize): Add support for singularize
It would be nice if pluralization also works for singularizing terms. For example,
Instead of `<dfn data-lt="bar">bars</dfn>`, we could write `<dfn>bars</dfn>` but still refer to it as `<a>bar</a>` or `<a>bars</a>`.

Presently it works opposite way only. i.e.
`<dfn>bar</dfn>` can be refereed as `<a>bar</a>` or `<a>bars</a>`, but `<dfn>bars</dfn>` cannot be referred as `<a>bar</a>`.

### PR DESCRIPTION
**Pull request by devanshbatra04**
Initial commit for adding singularize capability. I am pretty sure more refactor is required. Reviews appreciated @sidvishnoi @saschanaz @marcoscaceres 

Closes #1743

### CODE REVIEW

File: src/core/pluralize.js [hunk 1]
```diff
@@ -35,10 +41,20 @@ export function run(conf) {
...
         registerDefinition(dfn, uniquePlurals);
       }
+      if (singulars.length) {
+        const userDefinedPlurals = dfn.dataset.singulars
+          ? dfn.dataset.singulars.split("|")
+          : [];
+        const uniqueSingulars = [
+          ...new Set([...userDefinedPlurals, ...singulars]),
+        ];
+        dfn.dataset.singulars = uniqueSingulars.join("|");
```
Comment (sidvishnoi) [inline src/core/pluralize.js]: We would prefer to not add new data attributes, so lets merge `uniquePlurals` and `uniqueSingulars` into `uniquePlurals` itself (and add to `dfn.dataset.plurals`)

File: src/core/pluralize.js [hunk 2]
```diff
@@ -58,9 +74,16 @@ function getPluralizer() {
...
   // returns pluralized term if `text` needs pluralization, "" otherwise
-  return function pluralizeDfn(text) {
-    const normText = normalize(text.toLowerCase());
-    const plural = pluralOf(normText);
-    return links.has(plural) && !dfns.has(plural) ? plural : "";
+  return {
+    pluralizeDfn: text => {
+      const normText = normalize(text.toLowerCase());
+      const plural = pluralOf(normText);
+      return links.has(plural) && !dfns.has(plural) ? plural : "";
```
Comment (sidvishnoi) [inline src/core/pluralize.js]: Lets keep a single function and return an array of two items from it?

Comment (devanshbatra04) [reply src/core/pluralize.js]: Done, it seems I only ever needed to change 1 line.

File: src/core/pluralize.js (hunk 1 above)
Comment (devanshbatra04) [reply src/core/pluralize.js]: This isn't required if we use a single function as you suggested.

File: src/core/pluralize.js [hunk 3]
```diff
@@ -60,7 +64,9 @@ function getPluralizer() {
   // returns pluralized term if `text` needs pluralization, "" otherwise
```
Comment (sidvishnoi) [inline src/core/pluralize.js]:
```suggestion
  // returns pluralized/singularized term if `text` needs pluralization/singularization, "" otherwise
```

**Comment by sidvishnoi** [comment]
Brilliant! We just need to add tests for this now.

**Comment by devanshbatra04** [comment]
Yup, haven't ever written one. Will need to study up a bit and will send asap.

**Comment by devanshbatra04** [comment]
@sidvishnoi look forward to test reviews :)

File: tests/spec/core/pluralize-spec.js [hunk 4]
```diff
@@ -225,3 +208,49 @@ describe("Core - Pluralize", () => {
     expect(badLink.classList.contains("respec-offending-element")).toBeTruthy();
   });
 });
+
+it("doesn't mishandle <dfn> referring plural nouns without singulars", async () => {
```
Comment (sidvishnoi) [inline tests/spec/core/pluralize-spec.js]: Nit: Move the tests inside the `describe` block

File: tests/spec/core/pluralize-spec.js [hunk 5]
```diff
@@ -189,23 +189,6 @@ describe("Core - Pluralize", () => {
     ).toBeTruthy();
   });
 
-  it("doesn't add pluralization when no <a> references plural term", async () => {
```
Comment (sidvishnoi) [inline tests/spec/core/pluralize-spec.js]: any reason to remove this test?

Comment (sidvishnoi) [reply tests/spec/core/pluralize-spec.js]: Lets just add a test `"doesn't add singularization when no <a> references singular term"` and keep this here

Comment (devanshbatra04) [reply tests/spec/core/pluralize-spec.js]: Doesn't make sense. Essentially now either pluralization or singularization will be applied ( both being called Pluralization). We will always have one form of pluralization, which is ensured in the first test itself.

Comment (devanshbatra04) [reply tests/spec/core/pluralize-spec.js]: My Bad, I seem to have made a very bad error, I accidentally pushed it out of the describe block. Correcting.

Comment (sidvishnoi) [reply tests/spec/core/pluralize-spec.js]: We still need to ensure that both paths in code are tested.
So, lets try these:
``` html
<dfn data-lt="baz">bar</dfn> can be referenced as <a>baz</a> or <a>bar</a> 
so dataset.plurals should be empty as we're not referencing "bazs" or "bars"
```

``` html
<dfn data-lt="tables">chairs</dfn> can be referenced as <a>chairs</a> or <a>tables</a> 
so dataset.plurals should be empty as we're not referencing "table" or "chair".
```

But in above, if we respectively refer to plural/singular terms also, we should have them in `dataset.plurals`. Makes sense?

Comment (devanshbatra04) [reply tests/spec/core/pluralize-spec.js]: Yes, I did get your point. Both path are technically the same, so the same test works for both. but I will add the test as a good security measure. Really appreciate going out of your way to make things clear :)

**Comment by devanshbatra04** [comment]
@sidvishnoi could you please have a quick look here as well? just the test file :smile:

File: tests/spec/core/pluralize-spec.js [hunk 6]
```diff
@@ -206,6 +235,23 @@ describe("Core - Pluralize", () => {
...
+      </section>
+    `;
+    const ops = makeStandardOps({ pluralize: true }, body);
+    const doc = await makeRSDoc(ops);
+
+    const dfn = doc.querySelector("#section dfn");
+    expect(dfn.id).toEqual("dfn-tables");
+    const dfnlt = dfn.dataset.lt.split("|").sort();
+    const expectedDfnlt = "chairs|tables".split("|"); // no "table" here
+    expect(dfnlt).toEqual(expectedDfnlt);
```
Comment (sidvishnoi) [inline tests/spec/core/pluralize-spec.js]: We should also test for value of `dfn.dataset.plurals`.

File: tests/spec/core/pluralize-spec.js [hunk 7]
```diff
@@ -189,6 +189,35 @@ describe("Core - Pluralize", () => {
     ).toBeTruthy();
   });
 
+  it("doesn't mishandle <dfn> referring plural nouns without singulars", async () => {
```
Comment (sidvishnoi) [inline tests/spec/core/pluralize-spec.js]: I think we don't need this test as we don't need to test what `pluralize` package gives us. 
Unless `scissors` and `trousers` are names of potential APIs.

**Comment by devanshbatra04** [comment]
@sidvishnoi I changed the second test a bit, adding plural and singular terms in the same case, just for more scrutiny.

File: tests/spec/core/pluralize-spec.js [hunk 8]
```diff
@@ -3,40 +3,60 @@ describe("Core - Pluralize", () => {
...
       <section id="section">
-        <dfn data-lt="baz">bar</dfn> can be referenced
+        <dfn data-lt="baz">bars</dfn> can be referenced
         as <a>baz</a>
         or <a>bar</a>
         or <a>bars</a>
         or <a>bazs</a>
         but not as <a id="ignored-link" href="/PASS">bar</a>
       </section>
+      <section>
```
Comment (sidvishnoi) [inline tests/spec/core/pluralize-spec.js]: Nit: this section is redundant

File: tests/spec/core/pluralize-spec.js (hunk 7 above)
Comment (sidvishnoi) [reply tests/spec/core/pluralize-spec.js]: Lets remove this test

Comment (devanshbatra04) [reply tests/spec/core/pluralize-spec.js]: done. thanks.

File: tests/spec/core/pluralize-spec.js (hunk 8 above)
Comment (devanshbatra04) [reply tests/spec/core/pluralize-spec.js]: whoops.

**Pull request merged**